File: map.cpp
// MapManager.cpp (Lógica correspondente em C++)
#include "MapManager.h"
#include <cstdlib>
#include <ctime>

MapManager::MapManager() {
    std::srand(static_cast<unsigned int>(std::time(nullptr)));
}

void MapManager::GenerateNewMap() {
    grid.clear();
    grid.resize(ROWS, std::vector<Tile>(COLUMNS));

    for (int y = 0; y < ROWS; y++) {
        for (int x = 0; x < COLUMNS; x++) {
            TileType type = TileType::GRAMA;

            if (x < COLUMNS - 1) {
                int chance = std::rand() % 100;
                if (chance < 5) {
                    type = TileType::PEDRA;
                } else if (chance < 10) {
                    type = TileType::AGUA;
                } else if (chance < 15) {
                    type = TileType::MATO;
                }
            }

            grid[y][x] = { x, y, type, x == COLUMNS - 1 };
        }
    }
}

void MapManager::ChangeTileType(int x, int y, TileType newType) {
    if (x >= 0 && x < COLUMNS && y >= 0 && y < ROWS) {
        grid[y][x].type = newType;
    }
}

TileType MapManager::GetTileType(int x, int y) const {
    if (x >= 0 && x < COLUMNS && y >= 0 && y < ROWS) {
        return grid[y][x].type;
    }
    return TileType::PEDRA; // Se sair do mapa, considera como bloqueado
}

bool MapManager::IsPlantation(int x) const {
    return x == COLUMNS - 1;
}

File: map.cs
// MapManager.cs (Lógica correspondente em C#)
using System;

public enum TileType 
{
    Grama, Terra, Solo, Lama, Trilha, Pedra, Agua, Mato
}

public class Tile 
{
    public int X { get; private set; }
    public int Y { get; private set; }
    public TileType Type { get; set; }
    public bool IsPlantationZone => X == MapManager.COLUMNS - 1;

    public Tile(int x, int y, TileType type) 
    {
        X = x;
        Y = y;
        Type = type;
    }
}

public class MapManager 
{
    public const int COLUMNS = 20;
    public const int ROWS = 10;
    
    public Tile[,] Grid { get; private set; }
    private Random _random;

    public MapManager() 
    {
        _random = new Random();
    }

    public void GenerateNewMap() 
    {
        Grid = new Tile[COLUMNS, ROWS];

        for (int y = 0; y < ROWS; y++) 
        {
            for (int x = 0; x < COLUMNS; x++) 
            {
                TileType type = TileType.Grama;

                if (x < COLUMNS - 1) 
                {
                    double chance = _random.NextDouble();
                    if (chance < 0.05) type = TileType.Pedra;
                    else if (chance < 0.10) type = TileType.Agua;
                    else if (chance < 0.15) type = TileType.Mato;
                }

                Grid[x, y] = new Tile(x, y, type);
            }
        }
    }

    public void ChangeTileType(int x, int y, TileType newType) 
    {
        if (x >= 0 && x < COLUMNS && y >= 0 && y < ROWS) 
        {
            Grid[x, y].Type = newType;
        }
    }
}